src/praisonai/praisonai/cli/features/tools.py

```
@@ -624,17 +656,34 @@ def action_add(self, args: List[str], **kwargs) -> Dict[str, Any]:
+                filename = Path(stem).name
+                if not filename.endswith(".py"):
+                    filename += ".py"
                 dest = tools_dir / filename
-                
-                urllib.request.urlretrieve(raw_url, dest)
+
+                # Read with an explicit HTTPS request and a 1 MiB cap instead of
+                # urlretrieve (which follows redirects and writes unbounded output).
+                req = urllib.request.Request(
+                    raw_url, headers={"User-Agent": "praisonai-tools-add"}
+                )
```

Ruff (0.16.1)

[error] Audit URL open for permitted schemes. Allowing use of `file:` or custom schemes is often unexpected.

(S310)